### DISCUSSION TITLE
Why is this breaking in instant mode?

### DISCUSSION THREAD

**Discussion by facelessuser** [squidfunk/mkdocs-material#3037] [Q&A] [answered]
So recently, I've finally been experimenting with instant mode again, and it works great. It allowed me to create interactive playgrounds for Python over [here](https://facelessuser.github.io/coloraide/). Pyodide loads once for the site, and resposne is pretty good.

So I've been trying to get instant going in Pymdown Extensions. But there seems to be some weird issue related to instant. I had to switch over to KaTeX as MathJax is just too hard to get going in instant, but let's take both of them out of the equation. So in this example, I'm not running any math JS libraries. simply running Arithmatex in `generic` mode:

```
  - pymdownx.arithmatex:
      generic: true
```

So it is just spitting out `<div>` elements with escaped text. But only when instant is loaded, I get all sorts of errors in the console:
<img width="440" alt="Screen Shot 2021-09-19 at 9 01 35 AM" src="https://user-images.githubusercontent.com/1055125/133932351-ffc73448-fdea-4624-beeb-1a938740b1e3.png">
<img width="507" alt="Screen Shot 2021-09-19 at 9 01 47 AM" src="https://user-images.githubusercontent.com/1055125/133932354-10ad3be4-e634-451b-b01f-83682284ff6d.png">
<img width="521" alt="Screen Shot 2021-09-19 at 9 01 53 AM" src="https://user-images.githubusercontent.com/1055125/133932357-26543e47-133a-4b20-ab73-683d45cb3b66.png">
<img width="352" alt="Screen Shot 2021-09-19 at 9 03 10 AM" src="https://user-images.githubusercontent.com/1055125/133932395-170c05fd-ec02-4a38-b7ca-7d2b352d8ddb.png">

What is it doing and how can I make it stop? 🙃

**Comment by squidfunk** [comment, marked as answer]
Hmm, I need a reproducible example to debug this. I've just checked and instant loading works nicely with Arithmatex. I've also tested with the `\begin{align}` syntax, replacing the `$$` I'm using in the docs in [the MathJax guide](https://github.com/squidfunk/mkdocs-material/blob/master/docs/reference/mathjax.md):

```
\begin{align}
\operatorname{ker} f=\{g\in G:f(g)=e_{H}\}{\mbox{.}}
\end{align}
```

Everything renders as expected.

__Edit__:

> I'm not running any math JS libraries

If I remove the Math libraries, the content is just rendered as-is. No errors in the console.

**Reply by facelessuser** [reply]
Are you using generic mode?

**Reply by facelessuser** [reply]
I may have to start walking this backwards.

**Reply by facelessuser** [reply]
Keep in mind this occurs on navigating to a subsequent page, initial page load seems to not throw this error.

**Reply by squidfunk** [reply]
Yes, I'm using generic mode, and yes, I've navigated.

**Reply by facelessuser** [reply]
Okay, I found the culprit. There was some fenced block format that I used for MathJax. It would generate previews and wrap the actual content in script tags:

```
<script type="math/tex; mode=display">\begin{align}
    p(v_i=1|\mathbf{h}) & = \sigma\left(\sum_j w_{ij}h_j + b_i\right) \\
    p(h_j=1|\mathbf{v}) & = \sigma\left(\sum_i w_{ij}v_i + c_j\right)
\end{align}</script>
```

Pasting the above into a document and navigating to the page and pack will generate errors in "instant". It has nothing to do with Arithmatex directly or any of the math libraries. What is instant doing parsing scripts? Or what is it doing? Instant disabled, these are ignored.

**Reply by facelessuser** [reply]
The fix for me, if using KaTex, is to remove this MathJax style usage, but the above should not be breaking things.

**Reply by squidfunk** [reply]
It reevaluate scripts that are part of the Markdown after replacing the content:

https://github.com/squidfunk/mkdocs-material/blob/ea130d1f311440543dbd7de0877a54cfaf7b9160/src/assets/javascripts/integrations/instant/index.ts#L299-L325

The MathJax example in the docs defines the `script` inside Markdown. However, if you're putting it outside of Markdown, this snippet ensures it's reevaluated:

```
document$.subscribe(() => {
  MathJax.typesetPromise()
})
```

As listed [here](https://squidfunk.github.io/mkdocs-material/reference/mathjax/#arithmatex).

**Reply by facelessuser** [reply]
Should it ignore script types it knows nothing about?

I've got the failure to go away, and KaTex works great with instant (after a few CSS patches). MathJax no longer generates the errors after switching to generic, but having issues getting it to work in instant, that's a separate thing that I don't feel like getting into right now.

More interested in script parsing. I feel like it should be ignoring non-JS stuff. I don't need them anymore for what I'm doing, but I could see maybe do something similar in the future though.

**Reply by facelessuser** [reply]
Ah, MathJax works again 🙂! I had a typo. So, yeah, it might be good to consider not evaluating non-JS scripts.

**Reply by squidfunk** [reply]
Yeah, that's definitely a good idea. We could check for the absence of `type`, or for the match of a JS type. PR appreciated!

**Reply by facelessuser** [reply]
Okay, I may look into this sometime soon 🙂.

**Reply by facelessuser** [reply]
There seems to be other issues with instant parsing stuff it seems. I think it was parsing some UML SVG stuff.

```
Error: <tspan> attribute dy: Expected length, "NaN"
```

Just creates a lot of noise.

**Reply by squidfunk** [reply]
As ever, a reproducible case and I can look into it. Possibly related to the `DOMParser` which is used to extract the content from the fetched HTML.

**Reply by facelessuser** [reply]
I understand, it was more of an aside, I haven't gathered all the details, it was mainly why I opened a discussion. I was more digging for an understanding of what Material was doing so I could better debug.

My plan is to dig deeper. It seems the parser is making some assumptions. I will open an issue once I understand things better.

**Reply by facelessuser** [reply]
The last one wasn't Material's fault, it's actually down in Mermaid. Some buggy code. Thanks again for the help!
